pandas/core/missing.py

```diff
@@ -674,54 +674,53 @@ def interpolate_2d(
     return result
 
 
-def _cast_values_for_fillna(values, dtype: DtypeObj, has_mask: bool):
-    """
-    Cast values to a dtype that algos.pad and algos.backfill can handle.
-    """
-    # TODO: for int-dtypes we make a copy, but for everything else this
-    #  alters the values in-place.  Is this intentional?
+def _fillna_prep(values, mask=None):
+    # boilerplate for _pad_1d, _backfill_1d, _pad_2d, _backfill_2d
 
-    if needs_i8_conversion(dtype):
-        values = values.view(np.int64)
+    if mask is None:
+        mask = isna(values)
 
-    elif is_integer_dtype(values) and not has_mask:
-        # NB: this check needs to come after the datetime64 check above
-        # has_mask check to avoid casting i8 values that have already
-        #  been cast from PeriodDtype
-        values = ensure_float64(values)
+    mask = mask.view(np.uint8)
+    return mask
 
-    return values
 
+def _datetimelike_compat(func):
```

Comment:
I've just started (and I mean _just_) experimenting with having a numba version of the libs.

numpy datetime and timedeltas can be handled by numba. This means that _cast_values_for_fillna as removed here (but replaced with the decorator) and creating the view of the mask are unnecessary with numba. d449ca0bfc5c510bb59a8b3c6af225c9c3969039

i suspect that many of the i8 conversions in the main codebase could become redundant (presumably numba deals with this)

with this in mind, would it be worth considering moving any i8 conversions into _libs? This may produce a cleaner _libs api and  simplify the python code.

Reply:
oh. and back to the PR. can you type the decorator to preserve the signature of the decorated function (ignoring that values to the decorator will be superset of values in the function, since we don't yet have type parameters for numpy arrays)

Reply:
> I've just started (and I mean just) experimenting with having a numba version of the libs.

neat!

> with this in mind, would it be worth considering moving any i8 conversions into _libs? This may produce a cleaner _libs api and simplify the python code.

ill have to give that some thought.  one wrinkle that comes to mind is that cdef functions dont tend to play nicely with decorators

Reply:
may need a python interface layer in _libs, but early days yet since don't know how will handle pandas types that are defined as ctypes, without a performance hit.